src/Framework/Migrations/Contracts/Migration.php

```diff
@@ -16,9 +16,15 @@ abstract class Migration
     /**
      * Bootstrap migration logic.
      *
+     * @unreleased this method is now optional if migration class implements BatchMigration interface
      * @since 2.9.0
      */
-    abstract public function run();
+    public function run()
+    {
+        if ( ! is_subclass_of($this, BatchMigration::class)) {
+            throw new RuntimeException('run method is not defined.');
+        }
+    }
```

Comment:
Let's avoid this if we can. This sort of thing can't be caught by static analysis, so the only way to know if there's an issue is at runtime.

Reply:
If we switch everything to inheritance we shouldn't need to do this.

Reply:
I agree, resolved!